encoding/ccf/decode.go

```diff
@@ -0,0 +1,2039 @@
+func (d *Decoder) decodeEnum(typ *cadence.EnumType, types *cadenceTypeByCCFTypeID) (cadence.Value, error) {
+	fieldValues, err := d.decodeComposite(typ.Fields, types)
+	if err != nil {
+		return nil, err
+	}
+
+	v, err := cadence.NewMeteredEnum(
+		d.gauge,
+		len(fieldValues),
+		func() ([]cadence.Value, error) {
+			return fieldValues, nil
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// typ is already metered at creation.
+	return v.WithType(typ), nil
+}
+
+// decodePath decodes path-value as
+// language=CDDL
+// path-value = [
+//
+//	domain: uint,
+//	identifier: tstr,
+//
+// ]
+func (d *Decoder) decodePath() (cadence.Value, error) {
+	// Decode array head of length 2.
+	err := decodeCBORArrayWithKnownSize(d.dec, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	// Decode domain.
+	pathDomain, err := d.dec.DecodeUint64()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get domain identifier.
+	// Identifier() panics if pathDomain is invalid.
+	domain := common.PathDomain(pathDomain).Identifier()
+
+	common.UseMemory(d.gauge, common.MemoryUsage{
+		Kind: common.MemoryKindRawString,
+		// No need to add 1 to account for empty string: string is metered in Path struct.
+		Amount: uint64(len(domain)),
+	})
+
+	// Decode identifier.
+	identifier, err := d.dec.DecodeString()
+	if err != nil {
+		return nil, err
+	}
+
+	common.UseMemory(d.gauge, common.MemoryUsage{
+		Kind: common.MemoryKindRawString,
+		// No need to add 1 to account for empty string: string is metered in Path struct.
+		Amount: uint64(len(identifier)),
+	})
+
+	return cadence.NewMeteredPath(d.gauge, domain, identifier), nil
+}
+
+// decodeCapability decodes encoded capability-value as
+// language=CDDL
+// capability-value = [
+//
+//	address: address-value,
+//	path: path-value
+//
+// ]
+func (d *Decoder) decodeCapability(typ *cadence.CapabilityType, types *cadenceTypeByCCFTypeID) (cadence.Value, error) {
+	// typ can be different from runtime CapabilityType because borrow type can be nil.
+	// In this case, runtime type is encoded with the value (as tag content for tag CBORTagTypeAndValue).
+
+	// Check next encoded CBOR type.
+	nextType, err := d.dec.NextType()
+	if err != nil {
+		return nil, err
+	}
+
+	if nextType == cbor.TagType {
+		err := decodeCBORTagWithKnownNumber(d.dec, CBORTagTypeAndValue)
+		if err != nil {
+			return nil, fmt.Errorf("unexpected encoded value of Cadence type %s (%T): %s", typ.ID(), typ, err.Error())
+		}
+
+		// Decode ccf-type-and-value-message.
+		return d.decodeTypeAndValue(types)
+	}
+
+	// Decode array head of length 2.
+	err = decodeCBORArrayWithKnownSize(d.dec, 2)
+	if err != nil {
+		return nil, err
+	}
+
+	// Decode address.
+	address, err := d.decodeAddress()
+	if err != nil {
+		return nil, err
+	}
+
+	// Decode path.
+	path, err := d.decodePath()
+	if err != nil {
+		return nil, err
+	}
+
+	return cadence.NewMeteredStorageCapability(
+		d.gauge,
+		path.(cadence.Path),
+		address.(cadence.Address),
+		typ.BorrowType), nil
```

Comment:
```suggestion
		typ.BorrowType,
	), nil
```

Reply:
Done in commit 926661cbc672ec8711015c5a097586a81e2f3d2b.